src/toxcore/net_crypto/crypto_connection.rs

```diff
@@ -0,0 +1,393 @@
+            base_nonce: sent_nonce,
+            session_pk,
+            cookie_hash: cookie.hash(),
+            cookie: our_encrypted_cookie,
+        };
+        let handshake = CryptoHandshake::new(&dht_precomputed_key, handshake_payload, cookie);
+        let status = ConnectionStatus::NotConfirmed {
+            sent_nonce,
+            received_nonce,
+            peer_session_pk,
+            packet: StatusPacket::new(DhtPacket::CryptoHandshake(handshake))
+        };
+
+        CryptoConnection {
+            dht_precomputed_key,
+            peer_real_pk,
+            peer_dht_pk,
+            session_sk,
+            session_pk,
+            status,
+            udp_addr: None,
+            udp_received_time: None,
+            udp_send_attempt_time: None
+        }
+    }
+
+    /// Get `CookieRequest` or `CryptoHandshake` if it should be sent depending
+    /// on connection status and update sent counter
+    pub fn packet_to_send(&mut self) -> Option<DhtPacket> {
+        match self.status {
+            ConnectionStatus::CookieRequesting { ref mut packet, .. }
+            | ConnectionStatus::HandshakeSending { ref mut packet, .. }
+            | ConnectionStatus::NotConfirmed { ref mut packet, .. } => {
+                if packet.should_be_sent() {
+                    packet.num_sent += 1;
+                    packet.sent_time = SystemTime::now();
+                    Some(packet.packet.clone())
+                } else {
+                    None
+                }
+            },
+            ConnectionStatus::Established { .. } => None,
+        }
+    }
+
+    /// Check if this connection is timed out, i.e. we didn't receive expected
+    /// packet in time
+    pub fn is_timed_out(&self) -> bool {
+        match self.status {
+            ConnectionStatus::CookieRequesting { ref packet, .. }
+            | ConnectionStatus::HandshakeSending { ref packet, .. }
+            | ConnectionStatus::NotConfirmed { ref packet, .. } => packet.is_timed_out(),
+            ConnectionStatus::Established { .. } => false, // TODO: timeout?
+        }
+    }
+
+    /// Set time when last UDP packet was received to now
+    pub fn update_udp_received_time(&mut self) {
+        self.udp_received_time = Some(SystemTime::now())
+    }
+
+    /// Set time when we made an attempt to send UDP packet
+    pub fn update_udp_send_attempt_time(&mut self) {
+        self.udp_send_attempt_time = Some(SystemTime::now())
+    }
+
+    /// Check if we received the last UDP packet not later than 8 seconds ago
+    pub fn is_udp_alive(&self) -> bool {
+        self.udp_received_time
+            .and_then(|time| time.elapsed().ok())
+            .map(|duration| duration < Duration::from_secs(UDP_DIRECT_TIMEOUT))
+            .unwrap_or(false)
+    }
+
+    /// Check if we should send UDP packet regardless of whether UDP is dead or
+    /// alive. In this case we shouldn't rely on UDP only and send the same
+    /// packet via TCP relay
```

Comment:
then it should be renamed to `tcp_attempt_should_be_made`?

Reply:
No, there is 2 possible cases - UDP is dead or UDP is alive. If it's dead we will always send packet via TCP. But in some cases we can send packet via UDP as well. So the name is correct - it checks if we should send UDP packet together with TCP packet.